gst/nnstreamer/elements/gsttensor_trainer.c

```diff
@@ -471,6 +483,41 @@ gst_tensor_trainer_change_state (GstElement * element,
   return GST_STATE_CHANGE_FAILURE;
 }
 
+
+/**
+ * @brief Check if current epochs is complete, 
+ * tensor_trainer wait for one of epochs to complete before getting the results from the subplugin
+ */
+static gboolean
+gst_tensor_trainer_epochs_is_complete (GstTensorTrainer * trainer)
+{
+  int required_sample;
+
+  g_return_val_if_fail (trainer != NULL, FALSE);
+  g_return_val_if_fail (trainer->fw != NULL, FALSE);
+  g_return_val_if_fail (&trainer->prop != NULL, FALSE);
+
+  required_sample =
+      trainer->prop.num_training_samples + trainer->prop.num_validation_samples;
+  if (trainer->total_push_data_cnt % required_sample != 0)
+    return FALSE;
+
+  g_mutex_lock (&trainer->epoch_complete_lock);
+  if (trainer->is_epoch_complete) {
+    /* initialize */
+    trainer->is_epoch_complete = FALSE;
+    g_mutex_unlock (&trainer->epoch_complete_lock);
+    return TRUE;
```

Comment:
If we are at epoch N, can you assure that this "is_epoch_complete" is for epoch N, not epoch N-1?
(what if ```nnstreamer_trainer_notify_event()``` is called twice already?)

Reply:
I'm sure.
Actually, we don't need ```mutex lock``` in here.
This action occurs when the last sample required for ```n-1 epoch``` is entered into the subplugin.
The start of ```n epoch``` is when new data is entered to the chain function after that ```n-1 epoch```'s runstats is pushed to tensor_sink in the chain function.

Reply:
last of n-1 epoch -> tensor_trainer g_cond_wait,  -> subplugin call nnstreamer_trainer_notify_event -> tensor_trainer send signal -> tensor_trainer unlock -> n-1 epoch complete -> get model runstats -> create output tenors -> push gstbuffer -> start of n epoch ->

Reply:
Thank you